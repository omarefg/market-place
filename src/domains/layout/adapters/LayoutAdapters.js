import React, { createContext } from "react";

export const LayoutContext = createContext();

export function LayoutProvider(props) {
  const {
    components,
    children,
    higherProps: { useShoppingCart, useFeedback },
  } = props;
  const { Footer } = components;

  const {
    state: { cart, errors: shoppingCartErrors, loaders: shoppingCartLoaders },
    components: { BuyButton },
    handlers: { handleBuyClick },
    metadata: {
      loadersKeys: shoppingCartLoadersKeys,
      errorsKeys: shoppingCartErrorsKeys,
      assets: { shoppingCart },
    },
  } = useShoppingCart();

  const { components: feedbackComponents } = useFeedback();
  const value = {};

  return (
    <LayoutContext.Provider value={value}>
      {children}
      <Footer>
        <BuyButton
          onClick={handleBuyClick}
          total={cart.length}
          disabled={!cart.length}
          error={shoppingCartErrors[shoppingCartErrorsKeys.postBuy]}
          isLoading={Boolean(
            shoppingCartLoaders[shoppingCartLoadersKeys.postBuy]
          )}
          feedbackComponents={feedbackComponents}
          cartIcon={shoppingCart}
        />
      </Footer>
    </LayoutContext.Provider>
  );
}
