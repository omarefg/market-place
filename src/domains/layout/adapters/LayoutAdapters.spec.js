import React from "react";
import { render } from "@testing-library/react";
import { LayoutProvider } from "./LayoutAdapters";

describe("@domains/layout/adapters/LayoutProvider", () => {
  let wrapper, defaultProps;
  beforeEach(() => {
    defaultProps = {
      components: {
        Footer: () => <footer />,
      },
      children: <div />,
      higherProps: {
        useShoppingCart: jest.fn(() => ({
          state: { cart: [], errors: {}, loaders: {} },
          components: { BuyButton: () => <button /> },
          handlers: { handleBuyClick: jest.fn() },
          metadata: {
            loadersKeys: {},
            errorsKeys: {},
          },
        })),
        useFeedback: jest.fn(() => ({
          components: {},
        })),
      },
    };

    wrapper = (props = {}) => {
      const finalProps = { ...defaultProps, ...props };
      return render(<LayoutProvider {...finalProps} />);
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Given a valid call it must render without crashing", () => {
    //Arrange
    const { container } = wrapper();
    //Assert
    expect(container).toMatchSnapshot();
  });

  it("Given a valid call it must render two nodes, children and Footer", () => {
    //Arrange
    const { container } = wrapper();
    //Assert
    expect(container.childNodes.length).toBe(2);
  });

  it("Given valid props it must Call useFeedback and useShoppingCart", () => {
    //Arrange
    wrapper();
    //Assert
    expect(defaultProps.higherProps.useShoppingCart).toBeCalledTimes(1);
    expect(defaultProps.higherProps.useFeedback).toBeCalledTimes(1);
  });
});
