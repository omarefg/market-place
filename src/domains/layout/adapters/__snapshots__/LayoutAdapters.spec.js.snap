// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@domains/layout/adapters/LayoutProvider Given a valid call it must render without crashing 1`] = `
<div>
  <div />
  <footer />
</div>
`;
